File: C++/ps_baekjoon/2234/main.cpp
#include <bits/stdc++.h>
using namespace std;
using pos = pair<int, int>;

constexpr int mxNM = 50, NOT_VST = 0, dr[]{0, +1, 0, -1}, dc[]{+1, 0, -1, 0}, wall[]{1, 2, 4, 8};
int n, m, idx = 1, board[mxNM][mxNM], vst[mxNM][mxNM];
unordered_map<int, set<int>> graph;
unordered_map<int, int> room;
queue<pos> q;

int main()
{
    cin.tie(0)->sync_with_stdio(0);
    cin >> n >> m;
    for (int i = 0; i < m; ++i)
        for (int j = 0; j < n; ++j)
            cin >> board[i][j];
    for (int i = 0; i < m; ++i)
        for (int j = 0; j < n; ++j)
            if (!vst[i][j])
            {
                vst[i][j] = idx;
                int cnt = 1;
                q.emplace(i, j);
                while (!q.empty())
                {
                    auto [r, c] = q.front();
                    q.pop();
                    for (int direction = 0; direction < 4; ++direction)
                    {
                        int nr = r + dr[direction];
                        int nc = c + dc[direction];
                        if (!(0 <= nr && nr < m && 0 <= nc && nc < n))
                            continue;
                        if (board[nr][nc] & wall[direction])
                        {
                            if (vst[nr][nc] && vst[nr][nc] != idx)
                                graph[idx].emplace(vst[nr][nc]);
                            continue;
                        }
                        if (vst[nr][nc])
                            continue;
                        vst[nr][nc] = idx;
                        q.emplace(nr, nc);
                        ++cnt;
                    }
                }
                room[idx] = cnt;
                ++idx;
            }
    int max_room = 0;
    int max_room_sum = -1;
    for (const auto& e : graph)
    {
        auto [room1, connect] = e;
        for (const auto& room2 : connect)
            max_room_sum = max(max_room_sum, room[room1] + room[room2]);
    }
    for (const auto& e : room)
        max_room = max(max_room, e.second);
    cout << room.size() << '\n' << max_room << '\n' << max_room_sum;
    return 0;
}

File: Python/ps_baekjoon/2234/main.py
import sys
from queue import Queue
from collections import defaultdict

readline, write = (sys.stdin.readline, sys.stdout.write)
q = Queue()
room = {}

n, m = [int(i) for i in readline().split()]
board = [[int(i) for i in readline().split()][:n] for _ in range(m)]
graph = defaultdict(set)
vst = [[0] * (n) for _ in range(m)]
index = 1
for i in range(m):
    for j in range(n):
        if not vst[i][j]:
            vst[i][j] = index
            cnt = 1
            q.put((i, j))
            while not q.empty():
                r, c = q.get()
                for dr, dc, movable in zip([0, +1, 0, -1], [+1, 0, -1, 0], [1, 2, 4, 8]):
                    nr = r + dr
                    nc = c + dc
                    if not (0 <= nr < m and 0 <= nc < n):
                        continue
                    if board[nr][nc] & movable:
                        if vst[nr][nc] and vst[nr][nc] != index:
                            graph[index].add(vst[nr][nc])
                        continue
                    if vst[nr][nc]:
                        continue
                    vst[nr][nc] = index
                    q.put((nr, nc))
                    cnt += 1
            room[index] = cnt
            index += 1
maxi = max([room[t[0]] + room[t[1]] for t in [(key, val) for key in graph for val in graph[key]]])
write(f"{len(room)}\n{max(room.values())}\n{maxi}")